proto/vtctldata.proto

```diff
@@ -1860,12 +1860,25 @@ message ValidateKeyspaceResponse {
   map<string, ValidateShardResponse> results_by_shard = 2;
 }
 
+message ValidatePermissionsKeyspaceRequest {
+  string keyspace = 1;
+  // If you only want to validate a subset of the shards in the
+  // keyspace, then specify a list of shard names.
+  repeated string shards = 2;
+}
+
+message ValidatePermissionsKeyspaceResponse {
+}
```

Comment:
Does this not contain any results?

Reply:
Yeah, it returns an error if there's a validation failure: https://github.com/vitessio/vitess/blob/76abae45fe6c8881c5ec308d05d1b648d058fa79/go/vt/vtctl/vtctl.go#L3259-L3269

We could change that, of course, but here I'm primarily just porting these missing commands over.

Reply:
We have this as common practice to sometimes have empty responses, where it's "error or nothing" and with the potential of adding proto fields in the future without breaking gRPC signature.

Reply:
Guess this is more a question for a separate change perhaps. But given we have various inputs with different shards, wouldn't we want to reflect that in the output similar to the next command listed here?

Reply:
Valid point, worth extended discussion.

Reply:
"We" here would be whomever uses and consumes this command (certainly not me to date). This PR is about migrating the missing but still used/needed commands from the old client to the new, where maintaining "drop in compatibility" is the only concern. Why are we discussing this here? Is there some feature gap or request? If so, we should open a feature request that lays it out and someone can pick that up at a later date.

I'm actually leaning towards removing the endtoend test for it (`ValidatePermissions*`, as I did with `CopySchemaShard`) since I cannot find anything that actually uses it today (in vitess, vitess-operator, etc):
```
go/test/endtoend/sharded/sharded_keyspace_test.go:      err = clusterInstance.VtctlclientProcess.ExecuteCommand("ValidatePermissionsShard", fmt.Sprintf("%s/%s", keyspaceName, shard1.Name))
go/test/endtoend/sharded/sharded_keyspace_test.go:      err = clusterInstance.VtctlclientProcess.ExecuteCommand("ValidatePermissionsKeyspace", keyspaceName)
go/vt/vtctl/vtctl.go:                           name:   "ValidatePermissionsShard",
go/vt/vtctl/vtctl.go:                           method: commandValidatePermissionsShard,
go/vt/vtctl/vtctl.go:                           name:   "ValidatePermissionsKeyspace",
go/vt/vtctl/vtctl.go:                           method: commandValidatePermissionsKeyspace,
go/vt/vtctl/vtctl.go:func commandValidatePermissionsShard(ctx context.Context, wr *wrangler.Wrangler, subFlags *pflag.FlagSet, args []string) error {
go/vt/vtctl/vtctl.go:           return fmt.Errorf("the <keyspace/shard> argument is required for the ValidatePermissionsShard command")
go/vt/vtctl/vtctl.go:   return wr.ValidatePermissionsShard(ctx, keyspace, shard)
go/vt/vtctl/vtctl.go:func commandValidatePermissionsKeyspace(ctx context.Context, wr *wrangler.Wrangler, subFlags *pflag.FlagSet, args []string) error {
go/vt/vtctl/vtctl.go:           return fmt.Errorf("the <keyspace name> argument is required for the ValidatePermissionsKeyspace command")
go/vt/vtctl/vtctl.go:   return wr.ValidatePermissionsKeyspace(ctx, keyspace)
go/vt/vtctld/vtctld.go: actionRepo.RegisterKeyspaceAction("ValidatePermissionsKeyspace",
go/vt/vtctld/vtctld.go:                 return "", wr.ValidatePermissionsKeyspace(ctx, keyspace)
go/vt/vtctld/vtctld.go: actionRepo.RegisterShardAction("ValidatePermissionsShard",
go/vt/vtctld/vtctld.go:                 return "", wr.ValidatePermissionsShard(ctx, keyspace, shard)
go/vt/wrangler/permissions.go:// ValidatePermissionsShard validates all the permissions are the same
go/vt/wrangler/permissions.go:func (wr *Wrangler) ValidatePermissionsShard(ctx context.Context, keyspace, shard string) error {
go/vt/wrangler/permissions.go:// ValidatePermissionsKeyspace validates all the permissions are the same
go/vt/wrangler/permissions.go:func (wr *Wrangler) ValidatePermissionsKeyspace(ctx context.Context, keyspace string) error {
go/vt/wrangler/permissions.go:          return wr.ValidatePermissionsShard(ctx, keyspace, shards[0])
go/vt/wrangler/testlib/permissions_test.go:     // run ValidatePermissionsKeyspace, this should work
go/vt/wrangler/testlib/permissions_test.go:     if err := vp.Run([]string{"ValidatePermissionsKeyspace", primary.Tablet.Keyspace}); err != nil {
go/vt/wrangler/testlib/permissions_test.go:             t.Fatalf("ValidatePermissionsKeyspace failed: %v", err)
go/vt/wrangler/testlib/permissions_test.go:     // run ValidatePermissionsKeyspace again, this should now fail
go/vt/wrangler/testlib/permissions_test.go:     if err := vp.Run([]string{"ValidatePermissionsKeyspace", primary.Tablet.Keyspace}); err == nil || !strings.Contains(err.Error(), "has an extra user") {
go/vt/wrangler/testlib/permissions_test.go:             t.Fatalf("ValidatePermissionsKeyspace has unexpected err: %v", err)
```

So rather than spending time discussing how we could improve it... I think we should kill it off during the client transition as it no longer seems to serve a valid/used purpose as a client command. Unless you for some reason feel this is needed (and should be improved at a later time)?

Reply:
To provide some context, the `GetPermissions` client command simply returns the contents of the `mysql.user` table as a JSON doc ([tmc](https://github.com/vitessio/vitess/blob/a57ae939ec5dbddc02e9eac31e1519113e904eeb/go/vt/vttablet/tabletmanager/rpc_actions.go#L64-L67) -> [mysqlctl](https://github.com/vitessio/vitess/blob/a57ae939ec5dbddc02e9eac31e1519113e904eeb/go/vt/mysqlctl/permissions.go#L26-L52)):
```
❯ vtctldclient GetPermissions zone1-100
{
  "user_permissions": [
    {
      "host": "%",
      "user": "vt_repl",
      "password_checksum": "0",
      "privileges": {
        "Alter_priv": "N",
        "Alter_routine_priv": "N",
        "Create_priv": "N",
        "Create_role_priv": "N",
        "Create_routine_priv": "N",
        "Create_tablespace_priv": "N",
        "Create_tmp_table_priv": "N",
        "Create_user_priv": "N",
        "Create_view_priv": "N",
        "Delete_priv": "N",
        "Drop_priv": "N",
        "Drop_role_priv": "N",
        "Event_priv": "N",
        "Execute_priv": "N",
        "File_priv": "N",
        "Grant_priv": "N",
        "Index_priv": "N",
        "Insert_priv": "N",
        "Lock_tables_priv": "N",
        "Password_require_current": "",
        "Password_reuse_history": "",
        "Password_reuse_time": "",
        "Process_priv": "N",
        "References_priv": "N",
        "Reload_priv": "N",
        "Repl_client_priv": "N",
        "Repl_slave_priv": "Y",
        "Select_priv": "N",
        "Show_db_priv": "N",
        "Show_view_priv": "N",
        "Shutdown_priv": "N",
        "Super_priv": "N",
        "Trigger_priv": "N",
        "Update_priv": "N",
        "User_attributes": "",
        "account_locked": "N",
        "authentication_string": "",
        "max_connections": "0",
        "max_questions": "0",
        "max_updates": "0",
        "max_user_connections": "0",
        "password_expired": "N",
        "password_lifetime": "",
        "plugin": "mysql_native_password",
        "ssl_cipher": "",
        "ssl_type": "",
        "x509_issuer": "",
        "x509_subject": ""
      }
    },
...
```

The `ValidatePermissions*` commands then tells you if the output differs within a shard or a keyspace. I don't disagree that telling you what nodes differ and how would be useful (not in this PR, but later enhancements)... IF this is actually still useful in the broader sense today. I haven't seen anything to indicate that it it's still used or still useful.

Reply:
In looking for uses of the ValidatePermissons commands (I found none) I did find uses of CopySchemaShard. So I think I will have to port that over.

Reply:
One last note... FWIW the functions used for the `ValidatePermissions*` commands do include the details in the error message: https://github.com/vitessio/vitess/blob/a57ae939ec5dbddc02e9eac31e1519113e904eeb/go/vt/mysqlctl/tmutils/permissions.go#L164

So I'm going to put this comment thread to bed for now.  🙂